### PR TITLE
refactor: major refactor

### PR DESCRIPTION
**Pull request by slhck**
This adds new features like:

- Handling of multiple repos
- Interactive repo selection

Development improvements:

- Test suite using vitest
- Consistent formatting using biome

(Squashed from my original repo.)

To discuss:

- I moved to yarn since we internally exclusively use this, but I am not passionate about which package manager to use. You originally had both npm and pnpm; I'd prefer to settle for one.
- I removed the compiled dist files since I don't like to commit built assets; these should become irrelevant once this is published on NPM

Still TODO after PR is merged:

- Publish on NPM

### CODE REVIEW

File: src/cleaner.ts [hunk 1]
```diff
@@ -1,390 +1,535 @@
...
+			} else {
+				console.log(
+					"Non-interactive environment detected. Continuing without prompt.",
+				);
+			}
+		}
+
+		if (startIndex > endIndex) {
+			console.error("Start index is greater than end index");
+			process.exit(1);
```
Comment (PierreBeucher) [inline src/cleaner.ts]: Maybe throw an exception here instead ?

File: src/cleaner.ts [hunk 2]
```diff
@@ -1,390 +1,535 @@
...
+			for (const result of batchResults) {
+				if (result.status === "fulfilled") {
+					repositories.push(result.value);
+				} // we ignore errors for now (invalid IDs)
+			}
+		}
+
+		if (repositories.length === 0) {
+			console.warn("   No repositories found. Maybe try again with a different ID range?");
+			process.exit(1);
```
Comment (PierreBeucher) [inline src/cleaner.ts]: Maybe throw an exception here instead ?

File: src/cleaner.ts [hunk 3]
```diff
@@ -1,390 +1,535 @@
...
+		);
+
+		// warn user if parameters doesn't make sense or forgot to disable safety
+		if (
+			keepRegex === DEFAULT_KEEP_REGEX ||
+			deleteRegex === DEFAULT_DELETE_REGEX
+		) {
+			console.warn("");
+			console.warn(
+				`🤔 Hey, looks like you kept default keep and/or delete regex. By default, these regex won't mach anything for safety reasons.`,
```
Comment (PierreBeucher) [inline src/cleaner.ts:268]: Coming back to this I wonder if making these flags required instead might not be a good idea ?

File: README.md [hunk 4]
```diff
@@ -80,41 +109,95 @@ cat /tmp/tags.json | jq '.[] | .name + "\t" + .created_at ' -r
...
 - Filter out tags not matching keep regex, filter in tags matching delete regex
 - Get tag details with [_Get details of a registry repository tag_](https://docs.gitlab.com/ee/api/container_registry.html#get-details-of-a-registry-repository-tag) as List registry repository tags only provide tag name, not creation date
-- Filter in tags older than specified number of days 
+- Filter in tags older than specified number of days
+- Keep the most recent N tags if specified
 - Delete 'em !
 
+## Development
+
+Run `yarn dev -- <command>` to start the dev version of the CLI with a given command.
```
Comment (PierreBeucher) [inline README.md:195]:
```suggestion
Run `yarn install` to install dependencies. 

Run `yarn dev -- <command>` to start the dev version of the CLI with a given command.
```

**Review by PierreBeucher — APPROVED** [review]
That's quite awesome a PR, thanks :)

I read through the code, especially important logic parts (regex and date filters, deletion and important API calls) without finding obvious errors. Added a few suggestion, though they're not required - do as you seem fit. 

Also ran install/build/test locally smoothly, works great !

I see you are using mocked server to manage test. Once upon a time I had in mind to implement integration testing using a GitLab container provisioned via Pulumi (or Terraform) to setup test project with registry and images. 

I can add such tests in the repo to add a layer of small integration tests using a real GitLab instance - but I fear it might be overkill. What do you think ? (works great on my over repos, eg. [Novops](https://github.com/PierreBeucher/novops))

File: src/cleaner.ts (hunk 3 above)
Comment (slhck) [reply src/cleaner.ts:268]: I don't know, I think it's fine, as this requires the user to really think about what regexes they need, and is an additional safeguard.
I am not a big fan of "required options", if you understand what I mean.

**Comment by slhck** [comment]
Thanks for the review! I'll go ahead and merge now.

> I see you are using mocked server to manage test. Once upon a time I had in mind to implement integration testing using a GitLab container provisioned via Pulumi (or Terraform) to setup test project with registry and images.

This would, of course, be the most reliable option in terms of what API GitLab uses. That said, I think it's quite a massive workload to launch it? Would that also work locally? We'd then have to see the projects and registries with real images. So, as long as the GitLab API does not change, we can relatively safely mock the responses, and I'm leaning towards thinking that this is rather overkill 😉

File: src/cleaner.ts (hunk 3 above)
Comment (PierreBeucher) [reply src/cleaner.ts:268]: > this requires the user to really think about what regexes they need

Exactly what I had in mind initially. Let's keep it this way.

**Comment by PierreBeucher** [comment]
It's actually rather easy to do (since I know how to, mind) and it would work locally just fine. But yeah probably overkill for now !

**Pull request merged**